Importing the Dependencies

Laut Herr Resanovic:

- 3 Modelle

In [99]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
import xgboost as xgb
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score

## Problem Statement
The wine company wants to create new wine brand. The company wants to be able to infer the quality of the wine based on the several chemical components in wine. Therefore you should build the ML model which is capable to recongnise the quality of the wine based on the chemical properties. See wine data. 

## Data Import
Import the winedate.csv

In [22]:
df = pd.read_csv("winedata.csv")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

## Data Analysis
Initial decision which Models/Algorithms are going to be used as potential candidates for the final model. Elaborate why these models

Als erstes Modell um sich zu orientieren verwenden wir die lineare Regression da sie einen simplen Einstiegspunkt gibt und gut interpretierbar ist.

Wir wollen Classifier Modelle verwenden in dem wir jede Qualitätsstufe 1-10 als Klasse ansehen. Daher kommen für uns gängige Classifier wie:

KNN
RandomForest
GaussianNaiveBayes


in Frage. 

Diese Modelle kommen in Frage da sie in die oben genannte Kategorie der Klassifizierungs Modelle fallen die für diesen Anwendungsfall geeignet sind. Auch ein Entscheidungskriterium war die Bewertung mittles Accuracy, Precison, Recall und F1- Score.

## Data Preprocessing
Here you should prepare data accordingly

In [ ]:
#auf Nullwerte prüfen
print(df.isnull().sum())


fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64


In [ ]:
#auf Duplikate prüfen
df.duplicated().sum()

np.int64(240)

In [ ]:
#initial 240 Duplikate gefunden, diese gelöscht
df.drop_duplicates()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
5               7.4             0.660         0.00             1.8      0.075   
...             ...               ...          ...             ...        ...   
1593            6.8             0.620         0.08             1.9      0.068   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
5                    13.0                  40.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1593                 28.0                  38.0  0.99651  3.42       0.82   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
5         9.4        5  
...       ...      ...  
1593      9.5        6  
1594     10.5        5  
1595     11.2        6  
1597     10.2        5  
1598     11.0        6  

[1359 rows x 12 columns]

In [ ]:
#Zielwert vom Features trennen damit Zielwert nicht verfälscht wird da das Modell den Zielwert basierend auf den Features lernen soll
X = df.drop("quality", axis=1)
y = df["quality"]

X.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  
0      9.4  
1      9.8  
2      9.8  
3      9.8  
4      9.4

In [ ]:
#skalieren damit alle Werte auf der gleichen Skala sind und Ausreißer nicht so stark ins Gewicht fallen

#hätten die Trainings/Testsplits skalieren sollen nicht den ganten DataFrame :(

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

df_scaled = pd.DataFrame(X_scaled, columns=df.drop("quality", axis=1).columns)

df_scaled.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0      -0.528360          0.961877    -1.391472       -0.453218  -0.243707   
1      -0.298547          1.967442    -1.391472        0.043416   0.223875   
2      -0.298547          1.297065    -1.186070       -0.169427   0.096353   
3       1.654856         -1.384443     1.484154       -0.453218  -0.264960   
4      -0.528360          0.961877    -1.391472       -0.453218  -0.243707   
5      -0.528360          0.738418    -1.391472       -0.524166  -0.264960   
6      -0.241094          0.403229    -1.083370       -0.666062  -0.392483   
7      -0.585813          0.682553    -1.391472       -0.949853  -0.477498   
8      -0.298547          0.291499    -1.288771       -0.382271  -0.307468   
9      -0.470907         -0.155419     0.457144        2.526589  -0.349975   

   free sulfur dioxide  total sulfur dioxide   density        pH  sulphates  \
0            -0.466193             -0.379133  0.558274  1.288643  -0.579207   
1             0.872638              0.624363  0.028261 -0.719933   0.128950   
2            -0.083669              0.229047  0.134264 -0.331177  -0.048089   
3             0.107592              0.411500  0.664277 -0.979104  -0.461180   
4            -0.466193             -0.379133  0.558274  1.288643  -0.579207   
5            -0.274931             -0.196679  0.558274  1.288643  -0.579207   
6            -0.083669              0.381091 -0.183745 -0.072005  -1.169337   
7            -0.083669             -0.774449 -1.137769  0.511130  -1.110324   
8            -0.657454             -0.865676  0.028261  0.316751  -0.520193   
9             0.107592              1.688677  0.558274  0.251958   0.837107   

    alcohol  
0 -0.960246  
1 -0.584777  
2 -0.584777  
3 -0.584777  
4 -0.960246  
5 -0.960246  
6 -0.960246  
7 -0.397043  
8 -0.866379  
9  0.072294

## Training and build Models
These models will be used for the evaluation

In [86]:
# Splitten der Daten in Test und Train Split, Aufteilung hier 75% Training und 25% Testing. RandomState = 1
# Skalieren für Tree Based Models zwar nicht notwendig da sie robust sind, jedoch schadet es auch nicht und für Lineare Regression oder 
# Lasso wichtig da hier ein großer numerischer Ausreißer höher gewichtet wird, nicht weil er besonders Relevant ist sondern weil es einfach eine
# große Zahl ist
X_train0, X_test0, y_train0, y_test0 = train_test_split(X, y, test_size=0.25, random_state=3)

X_train0 = scaler.fit_transform(X_train0)
X_test0 = scaler.transform(X_test0)

In [87]:
#Random Forest initialisieren

model = RandomForestClassifier()
model.fit(X_train0, y_train0)

#vorhersage treffen
y_prediction0 = model.predict(X_test0)

#f1 score berechnen
f1_0 = f1_score(y_test0, y_prediction, average="weighted")
print(f"F1: {f1_0:.3f}")

F1: 0.384


In [88]:
#KNN initialisieren
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train0, y_train0)

#vorhersagen
y_prediction1 = knn.predict(X_test0)

#f1
f1_1 = f1_score(y_test0, y_prediction1, average="weighted")
print(f"F1: {f1_1:.3f}")

F1: 0.555


In [89]:
#NaiveBayes initialisieren
nb = GaussianNB()
nb.fit(X_train0, y_train0)

#vorhersagen
y_prediction2 = nb.predict(X_test0)

#f1
f1_2 = f1_score(y_test0, y_prediction2, average='weighted')
print(f"F1-Score (Naive Bayes): {f1_2:.3f}")

F1-Score (Naive Bayes): 0.564


In [92]:
# Zielwert-Encoder anwenden
le = LabelEncoder()
y_train0_encoded = le.fit_transform(y_train0)
y_test0_encoded = le.transform(y_test0)

#XGBoost
xgb_model = xgb.XGBClassifier(
    use_label_encoder=True, 
    eval_metric='mlogloss',
    max_depth = 6,
    n_estimators = 300,
)

xgb_model.fit(X_train0, y_train0_encoded)

#vorhersagen
y_prediction3 = xgb_model.predict(X_test0)

# Vorhersagen wieder in Originalwerte zurücktransformieren
y_prediction3_decoded = le.inverse_transform(y_prediction3)

#f1
f1_3 = f1_score(y_test0, y_prediction3_decoded, average='weighted')
print(f"F1-Score (XGBoost): {f1_3:.3f}")

print(classification_report(y_test0, y_prediction3_decoded))

/home/maninblacc/.local/lib/python3.12/site-packages/xgboost/training.py:183: UserWarning: [18:18:17] WARNING: /workspace/src/learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


F1-Score (XGBoost): 0.693
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         3
           4       0.25      0.07      0.11        14
           5       0.72      0.81      0.76       166
           6       0.70      0.70      0.70       166
           7       0.78      0.58      0.67        48
           8       0.25      0.67      0.36         3

    accuracy                           0.70       400
   macro avg       0.45      0.47      0.43       400
weighted avg       0.69      0.70      0.69       400



/home/maninblacc/.local/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/maninblacc/.local/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/maninblacc/.local/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


## Evaluation and selection of the model
Here multiple models should be evaluated and finaly selected. 

### Accuracy score
0.7025 - xgb
0.565 - knn
0.74 - rf
0.555 - nb

### Precision
0.6930955882352942 - xgb
0.5471185802940024 - knn
0.7095144235704752 - rf
0.581110922787194 - nb
### Recall

### F1 Score

In [100]:
accuracy_xgb = accuracy_score(y_test0, y_prediction3_decoded)
accuracy_knn = accuracy_score(y_test0, y_prediction1)
accuracy_rf = accuracy_score(y_test0, y_prediction0)
accuracy_nb = accuracy_score(y_test0, y_prediction2)

print(accuracy_xgb)
print(accuracy_knn)
print(accuracy_rf)
print(accuracy_nb)

0.7025
0.565
0.74
0.555


In [102]:
precision_xgb = precision_score(y_test0, y_prediction3_decoded, average='weighted')
precision_knn = precision_score(y_test0, y_prediction1, average='weighted')
precision_rf = precision_score(y_test0, y_prediction0, average='weighted')
precision_nb = precision_score(y_test0, y_prediction2, average='weighted')

print(precision_xgb)
print(precision_knn)
print(precision_rf)
print(precision_nb)

0.6930955882352942
0.5471185802940024
0.7095144235704752
0.581110922787194


/home/maninblacc/.local/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/maninblacc/.local/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/maninblacc/.local/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


## Conclusion and reflection
Please reflect on the project and provide the conclusion

## Test the model with some data
Create your own new data representing wine and test the model